### PR TITLE
Add changelog generation to the versioner tool

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Updates to existing modules will generate a changelog in the module
directory; new modules won't contain a changelog.
Refactor package report generation to its own package.

- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] The PR targets the `latest` branch.
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] Apache v2 license headers are included in each file.
 
[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/master/CHANGELOG.md

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
## Breaking Changes

### Removed Funcs

1. ResourceProperties.MarshalJSON() ([]byte,error)

## Struct Changes

### Removed Structs

1. LinkedServerList
1. ResourceProperties

### Removed Struct Fields

1. CreateParameters.ID
1. CreateParameters.Name
1. CreateParameters.Type
1. PatchSchedule.Location
1. ResourceType.*ResourceProperties
1. UpdateProperties.StaticIP
1. UpdateProperties.SubnetID

## Signature Changes

### Funcs

1. FirewallRulesClient.CreateOrUpdate
	- Params
		- From: context.Context,string,string,string,FirewallRule
		- To: context.Context,string,string,string,FirewallRuleCreateParameters
1. FirewallRulesClient.CreateOrUpdatePreparer
	- Params
		- From: context.Context,string,string,string,FirewallRule
		- To: context.Context,string,string,string,FirewallRuleCreateParameters
1. LinkedServerClient.List
	- Returns
		- From: LinkedServerWithPropertiesList,error
		- To: LinkedServerWithPropertiesListPage,error

## New Content

### New Constants

1. ProvisioningState.Creating
1. ProvisioningState.Deleting
1. ProvisioningState.Disabled
1. ProvisioningState.Failed
1. ProvisioningState.Linking
1. ProvisioningState.Provisioning
1. ProvisioningState.RecoveringScaleFailure
1. ProvisioningState.Scaling
1. ProvisioningState.Succeeded
1. ProvisioningState.Unlinking
1. ProvisioningState.Unprovisioning
1. ProvisioningState.Updating
1. TLSVersion.OneFullStopOne
1. TLSVersion.OneFullStopTwo
1. TLSVersion.OneFullStopZero

### New Funcs

1. *FirewallRuleCreateParameters.UnmarshalJSON([]byte) error
1. *LinkedServerWithPropertiesListIterator.Next() error
1. *LinkedServerWithPropertiesListIterator.NextWithContext(context.Context) error
1. *LinkedServerWithPropertiesListPage.Next() error
1. *LinkedServerWithPropertiesListPage.NextWithContext(context.Context) error
1. *PatchScheduleListResultIterator.Next() error
1. *PatchScheduleListResultIterator.NextWithContext(context.Context) error
1. *PatchScheduleListResultPage.Next() error
1. *PatchScheduleListResultPage.NextWithContext(context.Context) error
1. Client.CheckNameAvailability(context.Context,CheckNameAvailabilityParameters) (autorest.Response,error)
1. Client.CheckNameAvailabilityPreparer(context.Context,CheckNameAvailabilityParameters) (*http.Request,error)
1. Client.CheckNameAvailabilityResponder(*http.Response) (autorest.Response,error)
1. Client.CheckNameAvailabilitySender(*http.Request) (*http.Response,error)
1. Client.ListUpgradeNotifications(context.Context,string,string,float64) (NotificationListResponse,error)
1. Client.ListUpgradeNotificationsPreparer(context.Context,string,string,float64) (*http.Request,error)
1. Client.ListUpgradeNotificationsResponder(*http.Response) (NotificationListResponse,error)
1. Client.ListUpgradeNotificationsSender(*http.Request) (*http.Response,error)
1. CommonProperties.MarshalJSON() ([]byte,error)
1. FirewallRuleCreateParameters.MarshalJSON() ([]byte,error)
1. LinkedServerClient.ListComplete(context.Context,string,string) (LinkedServerWithPropertiesListIterator,error)
1. LinkedServerWithPropertiesList.IsEmpty() bool
1. LinkedServerWithPropertiesListIterator.NotDone() bool
1. LinkedServerWithPropertiesListIterator.Response() LinkedServerWithPropertiesList
1. LinkedServerWithPropertiesListIterator.Value() LinkedServerWithProperties
1. LinkedServerWithPropertiesListPage.NotDone() bool
1. LinkedServerWithPropertiesListPage.Response() LinkedServerWithPropertiesList
1. LinkedServerWithPropertiesListPage.Values() []LinkedServerWithProperties
1. NewLinkedServerWithPropertiesListIterator(LinkedServerWithPropertiesListPage) LinkedServerWithPropertiesListIterator
1. NewLinkedServerWithPropertiesListPage(func(context.Context, LinkedServerWithPropertiesList) (LinkedServerWithPropertiesList, error)) LinkedServerWithPropertiesListPage
1. NewPatchScheduleListResultIterator(PatchScheduleListResultPage) PatchScheduleListResultIterator
1. NewPatchScheduleListResultPage(func(context.Context, PatchScheduleListResult) (PatchScheduleListResult, error)) PatchScheduleListResultPage
1. PatchScheduleListResult.IsEmpty() bool
1. PatchScheduleListResultIterator.NotDone() bool
1. PatchScheduleListResultIterator.Response() PatchScheduleListResult
1. PatchScheduleListResultIterator.Value() PatchSchedule
1. PatchScheduleListResultPage.NotDone() bool
1. PatchScheduleListResultPage.Response() PatchScheduleListResult
1. PatchScheduleListResultPage.Values() []PatchSchedule
1. PatchSchedulesClient.ListByRedisResource(context.Context,string,string) (PatchScheduleListResultPage,error)
1. PatchSchedulesClient.ListByRedisResourceComplete(context.Context,string,string) (PatchScheduleListResultIterator,error)
1. PatchSchedulesClient.ListByRedisResourcePreparer(context.Context,string,string) (*http.Request,error)
1. PatchSchedulesClient.ListByRedisResourceResponder(*http.Response) (PatchScheduleListResult,error)
1. PatchSchedulesClient.ListByRedisResourceSender(*http.Request) (*http.Response,error)
1. PossibleProvisioningStateValues() []ProvisioningState
1. PossibleTLSVersionValues() []TLSVersion
1. UpgradeNotification.MarshalJSON() ([]byte,error)

## Struct Changes

### New Structs

1. CheckNameAvailabilityParameters
1. CommonProperties
1. FirewallRuleCreateParameters
1. LinkedServerWithPropertiesListIterator
1. LinkedServerWithPropertiesListPage
1. NotificationListResponse
1. PatchScheduleListResult
1. PatchScheduleListResultIterator
1. PatchScheduleListResultPage
1. UpgradeNotification

### New Struct Fields

1. CreateParameters.Zones
1. CreateProperties.MinimumTLSVersion
1. LinkedServerWithPropertiesList.NextLink
1. Properties.AccessKeys
1. Properties.HostName
1. Properties.LinkedServers
1. Properties.MinimumTLSVersion
1. Properties.Port
1. Properties.ProvisioningState
1. Properties.RedisVersion
1. Properties.Sku
1. Properties.SslPort
1. ResourceType.*Properties
1. ResourceType.Zones
1. UpdateProperties.MinimumTLSVersion

**Comment by jhendrixMSFT** [comment]
@tombuildsstuff Here's the updated diff tool to generate markdown, see example I pasted earlier.

File: tools/versioner/cmd/root.go [hunk 1]
```diff
@@ -233,8 +231,23 @@ func updateGoModVer(goMod io.ReadWriteSeeker, newVer string) error {
...
+	const changeLogName = "CHANGELOG.md"
+	rpt := mod.GenerateReport()
+	log, err := os.Create(filepath.Join(stage, changeLogName))
+	if err != nil {
+		return fmt.Errorf("failed to create %s: %v", changeLogName, err)
+	}
+	defer log.Close()
+	if rpt.IsEmpty() {
+		_, err = log.WriteString("No changes to exported content compared to the previous release.\n")
+		return err
```
Comment (vladbarosan) [inline tools/versioner/cmd/root.go]: in what case would this happen ? and do we want it to be an error ?

File: tools/versioner/cmd/root_test.go [hunk 2]
```diff
@@ -344,6 +365,9 @@ go 1.12
 	if !strings.EqualFold(string(b), after) {
 		t.Fatalf("bad go.mod file, expected '%s' got '%s'", after, string(b))
 	}
+	if !fileExists(filepath.Join("../testdata/scenariob/foo/v2", "CHANGELOG.md")) {
+		t.Fatal("expected changelog in scenariob")
```
Comment (vladbarosan) [inline tools/versioner/cmd/root_test.go]: >scenariob [](start = 33, length = 9)

hmm what is scenariob ? would it be possible to give them more semanticNames ?

File: tools/versioner/cmd/root.go (hunk 1 above)
Comment (jhendrixMSFT) [reply tools/versioner/cmd/root.go]: Can't be an error as this is legit.  Most common case is documentation changes (i.e. comments).  Next case would be a generator change the modifies code bodies.

File: tools/versioner/cmd/root_test.go (hunk 2 above)
Comment (jhendrixMSFT) [reply tools/versioner/cmd/root_test.go]: I added comment text in each foo.go file that describes the scenario, figured that might be better than trying to get create with the names.  I will double-check that I didn't miss one.

**Pull request merged**